### PR TITLE
Implements translation of the legacy CUDA code to HIP

### PR DESCRIPTION
**Pull request by jakurzak**
## Proposed changes

The legacy CUDA code (QMC_CUDA) can be compiled for AMD GPUs using HIP.
This is done by pointing to hipcc as the compiler, and using the `-DQMC_CUDA2HIP` CMake flag in addition to the `-DQMC_CUDA` flag, i.e.:

`-DCMAKE_C_COMPILER=hipcc -DCMAKE_CXX_COMPILER=hipcc -DQMC_CUDA=1 -DQMC_CUDA2HIP=1`

The changes have no impact on the CUDA code.
The translation is accomplished by including a header file with translations of CUDA API to HIP API, e.g.:

```
#define cudaMalloc                  hipMalloc
#define cudaMallocArray             hipMallocArray
#define cudaMallocManaged           hipMallocManaged
#define cudaMemAdvise               hipMemAdvise
...
```

## What systems has this change been tested on?

- Radeon VII, ROCm 4.1
- MI100, ROCm 4.2
- V100, CUDA 11.2

### CODE REVIEW

**Comment by prckent** [comment]
Thanks Jakub! A surpringly small change. We'll have a play. I expect a little discussion around the build flags plus clearly src/cuda2hip.h should be in src/Platforms/something.

**Comment by prckent** [comment]
Q. Have you verified this works using (say) openmpi wrappers configured to use hipcc? or Cray compiler wrappers?

**Comment by ye-luo** [comment]
```
/home/yeluo/opt/qmcpack/src/QMCWaveFunctions/EinsplineSetCuda.cpp:790:30: error: use of undeclared identifier 'cudaIpcMemLazyEnablePeerAccess'
                             cudaIpcMemLazyEnablePeerAccess);
```

**Comment by jakurzak** [comment]
> Q. Have you verified this works using (say) openmpi wrappers configured to use hipcc? or Cray compiler wrappers?

**Comment by jakurzak** [comment]
Yes, it is surprisingly non-intrusive.
I ran with MPI compiler wrappers, but not recently.
I will test with OpenMPI wrappers and Cray wrappers.
Thumbs up to splitting the header file. Please let me know the desired location.
I have not seen the cudaIpcMemLazyEnablePeerAccess error.

**Comment by jakurzak** [comment]
I am sorry, MPI compilation issues were my omissions.
It should be fixed now - tested with OpenMPI and MPICH.

**Comment by prckent** [comment]
(Trying to compile this but running into rocm dependency issues and other newb problems.)

File: CMakeLists.txt [hunk 1]
```diff
@@ -723,35 +723,60 @@ ENDIF()
...
+    INCLUDE_DIRECTORIES(${CUDA_INCLUDE_DIRS})
+    SET(HAVE_CUDA 1)
+    MESSAGE("   CUDA_NVCC_FLAGS=${CUDA_NVCC_FLAGS}")
+  ELSE()
+    SET(CMAKE_MODULE_PATH "/opt/rocm/hip/cmake;${CMAKE_MODULE_PATH}")
+    SET(CMAKE_MODULE_PATH "$ENV{ROCM_PATH}/hip/cmake;${CMAKE_MODULE_PATH}")
+    FIND_PACKAGE(HIP REQUIRED)
+    INCLUDE_DIRECTORIES("/opt/rocm/include")
+    INCLUDE_DIRECTORIES("$ENV{ROCM_PATH}/include")
+    ADD_DEFINITIONS(-D__HIP_PLATFORM_HCC__=)
```
Comment (ye-luo) [inline CMakeLists.txt:762]: Based on rocm `__HIP_PLATFORM_AMD__` is up-to-date flag.

Comment (jakurzak) [reply CMakeLists.txt]: Thank you for pointing out.
Just committed the change to `__HIP_PLATFORM_AMD__`.

File: CMakeLists.txt [hunk 2]
```diff
@@ -723,35 +723,61 @@ ENDIF()
...
+    INCLUDE_DIRECTORIES(${CUDA_INCLUDE_DIRS})
+    SET(HAVE_CUDA 1)
+    MESSAGE("   CUDA_NVCC_FLAGS=${CUDA_NVCC_FLAGS}")
+  ELSE()
+    SET(CMAKE_MODULE_PATH "/opt/rocm/hip/cmake;${CMAKE_MODULE_PATH}")
+    SET(CMAKE_MODULE_PATH "$ENV{ROCM_PATH}/hip/cmake;${CMAKE_MODULE_PATH}")
+    FIND_PACKAGE(HIP REQUIRED)
+    INCLUDE_DIRECTORIES("/opt/rocm/include")
+    INCLUDE_DIRECTORIES("$ENV{ROCM_PATH}/include")
+    ADD_DEFINITIONS(-D__HIP_PLATFORM_AMD__=)
```
Comment (ye-luo) [inline CMakeLists.txt:762]: I don't think `=` is needed.

File: CMakeLists.txt [hunk 3]
```diff
@@ -723,35 +723,61 @@ ENDIF()
...
+      SET(CUDA_NVCC_FLAGS "-arch=${CUDA_ARCH};${CUDA_NVCC_FLAGS}")
+    endif(CUDA_NVCC_FLAGS MATCHES "arch")
+    INCLUDE_DIRECTORIES(${CUDA_INCLUDE_DIRS})
+    SET(HAVE_CUDA 1)
+    MESSAGE("   CUDA_NVCC_FLAGS=${CUDA_NVCC_FLAGS}")
+  ELSE()
+    SET(CMAKE_MODULE_PATH "/opt/rocm/hip/cmake;${CMAKE_MODULE_PATH}")
+    SET(CMAKE_MODULE_PATH "$ENV{ROCM_PATH}/hip/cmake;${CMAKE_MODULE_PATH}")
+    FIND_PACKAGE(HIP REQUIRED)
+    INCLUDE_DIRECTORIES("/opt/rocm/include")
```
Comment (ye-luo) [inline CMakeLists.txt:760]: hard-coded path is bad. FIND_PACKAGE(HIP REQUIRED) should have setup some variables and then use them in INCLUDE_DIRECTORIES

File: CMakeLists.txt [hunk 4]
```diff
@@ -723,35 +723,61 @@ ENDIF()
...
+    INCLUDE_DIRECTORIES("$ENV{ROCM_PATH}/include")
+    ADD_DEFINITIONS(-D__HIP_PLATFORM_AMD__=)
+    # hipblas
+    LINK_DIRECTORIES("/opt/rocm/lib")
```
Comment (ye-luo) [inline CMakeLists.txt:764]: hard-coded path is bad. Try find_package hipblas and use variables comes out of it.

File: CMakeLists.txt [hunk 5]
```diff
@@ -723,35 +723,61 @@ ENDIF()
...
+    LINK_DIRECTORIES("$ENV{ROCM_PATH}/lib")
+    LINK_DIRECTORIES("/opt/rocm/hipblas/lib")
+    LINK_DIRECTORIES("$ENV{ROCM_PATH}/hipblas/lib")
+    INCLUDE_DIRECTORIES("/opt/rocm/hipblas/include")
+    INCLUDE_DIRECTORIES("$ENV{ROCM_PATH}/hipblas/include")
+    LINK_LIBRARIES(amdhip64 hipblas)
+    # architecture flags
+    SET(HIP_HIPCC_FLAGS "${HIP_HIPCC_FLAGS} -fPIC")
+    SET(HIP_HIPCC_FLAGS "${HIP_HIPCC_FLAGS} -fopenmp")
+    SET(HIP_HIPCC_FLAGS "${HIP_HIPCC_FLAGS} --amdgpu-target=gfx906,gfx908")
```
Comment (ye-luo) [inline CMakeLists.txt:774]: Set up a cached variable HIP_ARCH and default to `gfx906,gfx908`

File: src/cuda2hip.h [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+#define cudaStreamSynchronize       hipStreamSynchronize
+#define cudaStreamWaitEvent         hipStreamWaitEvent
+#define cudaSuccess                 hipSuccess
+#define cuDoubleComplex             hipDoubleComplex
+#define make_cuComplex              make_hipComplex
+#define make_cuDoubleComplex        make_hipDoubleComplex
+
+#define cudaDeviceSetLimit(limit, falue) ;
+
+#include <hipblas.h>
```
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/cuda2hip.h:106]: `src/Platforms/ROCm/hipBLAS.hpp`

**Comment by ye-luo** [comment]
Full precision build.
```
cmake -DCMAKE_C_COMPILER=hipcc -DCMAKE_CXX_COMPILER=hipcc -DQMC_MPI=0 -DQMC_CUDA=1 -DQMC_CUDA2HIP=1 -DQMC_MIXED_PRECISION=OFF ..
```
running tests.
```
$ ctest -R deter -j32
...
84% tests passed, 33 tests failed out of 210

Label Time Summary:
QMCPACK                     =  46.24 sec*proc (15 tests)
QMCPACK-checking-results    =   5.27 sec*proc (164 tests)
deterministic               =  67.31 sec*proc (210 tests)
quality_unknown             =  67.31 sec*proc (210 tests)
unit                        =  15.80 sec*proc (31 tests)

Total Test time (real) =  62.13 sec

The following tests FAILED:
	522 - deterministic-diamondC_2x1x1_pp-vmc_sdj-1-1-kinetic (Failed)
	523 - deterministic-diamondC_2x1x1_pp-vmc_sdj-1-1-totenergy (Failed)
	524 - deterministic-diamondC_2x1x1_pp-vmc_sdj-1-1-eeenergy (Failed)
	526 - deterministic-diamondC_2x1x1_pp-vmc_sdj-1-1-potential (Failed)
	528 - deterministic-diamondC_2x1x1_pp-vmc_sdj-1-1-localecp (Failed)
	529 - deterministic-diamondC_2x1x1_pp-vmc_sdj-1-1-nonlocalecp (Failed)
	530 - deterministic-diamondC_2x1x1_pp-vmc_sdj-1-1-flux (Failed)
	532 - deterministic-diamondC_2x1x1_pp-vmc_sdj_excited-1-1-kinetic (Failed)
	533 - deterministic-diamondC_2x1x1_pp-vmc_sdj_excited-1-1-totenergy (Failed)
	534 - deterministic-diamondC_2x1x1_pp-vmc_sdj_excited-1-1-eeenergy (Failed)
	536 - deterministic-diamondC_2x1x1_pp-vmc_sdj_excited-1-1-potential (Failed)
	538 - deterministic-diamondC_2x1x1_pp-vmc_sdj_excited-1-1-localecp (Failed)
	539 - deterministic-diamondC_2x1x1_pp-vmc_sdj_excited-1-1-nonlocalecp (Failed)
	540 - deterministic-diamondC_2x1x1_pp-vmc_sdj_excited-1-1-flux (Failed)
	542 - deterministic-diamondC_2x1x1_pp-delayed_update-vmc_sdj-1-1-kinetic (Failed)
	543 - deterministic-diamondC_2x1x1_pp-delayed_update-vmc_sdj-1-1-totenergy (Failed)
	544 - deterministic-diamondC_2x1x1_pp-delayed_update-vmc_sdj-1-1-eeenergy (Failed)
	546 - deterministic-diamondC_2x1x1_pp-delayed_update-vmc_sdj-1-1-potential (Failed)
	548 - deterministic-diamondC_2x1x1_pp-delayed_update-vmc_sdj-1-1-localecp (Failed)
	549 - deterministic-diamondC_2x1x1_pp-delayed_update-vmc_sdj-1-1-nonlocalecp (Failed)
	550 - deterministic-diamondC_2x1x1_pp-delayed_update-vmc_sdj-1-1-flux (Failed)
	552 - deterministic-diamondC_2x1x1_pp-dmc_sdj-1-1-kinetic (Failed)
	553 - deterministic-diamondC_2x1x1_pp-dmc_sdj-1-1-totenergy (Failed)
	554 - deterministic-diamondC_2x1x1_pp-dmc_sdj-1-1-eeenergy (Failed)
	556 - deterministic-diamondC_2x1x1_pp-dmc_sdj-1-1-potential (Failed)
	558 - deterministic-diamondC_2x1x1_pp-dmc_sdj-1-1-localecp (Failed)
	559 - deterministic-diamondC_2x1x1_pp-dmc_sdj-1-1-nonlocalecp (Failed)
	561 - deterministic-diamondC_2x1x1_pp-dmc_sdj_excited-1-1-kinetic (Failed)
	562 - deterministic-diamondC_2x1x1_pp-dmc_sdj_excited-1-1-totenergy (Failed)
	563 - deterministic-diamondC_2x1x1_pp-dmc_sdj_excited-1-1-eeenergy (Failed)
	565 - deterministic-diamondC_2x1x1_pp-dmc_sdj_excited-1-1-potential (Failed)
	567 - deterministic-diamondC_2x1x1_pp-dmc_sdj_excited-1-1-localecp (Failed)
	568 - deterministic-diamondC_2x1x1_pp-dmc_sdj_excited-1-1-nonlocalecp (Failed)
Errors while running CTest
```

**Comment by jakurzak** [comment]
So, yes, I am aware that about 15% of the deterministic tests fail.
Since the port is based purely on renaming, this does not affect the CUDA code, only the HIP port.
Hopefully, only some small senamntic difference between HIP and CUDA.
Can I get assistance with finding the culprit?
Is it an option to accept the PR to develop with this issue?
Looks like it's the 2x1x1 that have a problem.

**Comment by prckent** [comment]
On accepting this PR:

1. I want to be able to compile and run this *somewhere* first. I have not yet been successful with my own rocm install on our own hardware. Fortunately we get access to spock@OLCF shortly and this should suffice.

2. We need to at least understand the severity of the problems if not fix them. That 2x1x1 deterministic tests are failing could indicate problems with a phase factor, complex data, sincos, exponentials... etc. somewhere relating to the splines. It is likely a real bug and not a precision or numerical rounding issue. However we need some evidence to narrow things down. When we run all the `short` tests the failure pattern and usable statistics will help. At that point we either create an issue identifying the problem or simply fix the problem.

**Comment by jakurzak** [comment]
The solution to this problem with CMake:
```
-- Check for working C compiler: /usr/bin/hipcc - broken
...
    error: invalid argument '-std=c++11' not allowed with 'C'
```
is to add the following CMake flags:
```
    -DCMAKE_C_STANDARD=99 \
    -DCMAKE_CXX_STANDARD=14 \
```

File: src/cuda2hip.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (jakurzak) [reply src/cuda2hip.h]: done

File: CMakeLists.txt (hunk 2 above)
Comment (jakurzak) [reply CMakeLists.txt]: fixed

File: CMakeLists.txt (hunk 3 above)
Comment (jakurzak) [reply CMakeLists.txt]: Without the path in `CMAKE_MODULE_PATH` CMake does not find it.
I tried to make it clean:
```
	IF(NOT DEFINED ROCM_ROOT)
        SET(ROCM_ROOT "/opt/rocm")
    ENDIF()
	LIST(APPEND CMAKE_MODULE_PATH ${ROCM_ROOT}/hip/cmake)
```

File: CMakeLists.txt (hunk 4 above)
Comment (jakurzak) [reply CMakeLists.txt]: Fixed.
No more hardcoded path to hipBLAS.

File: CMakeLists.txt (hunk 5 above)
Comment (jakurzak) [reply CMakeLists.txt]: done

**Comment by prckent** [comment]
Small update to say that I have this compiled and running on nitrogen (Radeon VII, ROCm 4.2). I am rerunning the tests with longer timeouts than I set initially. I am looking for any patterns in the short tests to help identify problems. Generally results were consistent with the above. I'll note that the _untuned_ performance tests for NiO looked to be running at roughly the correct speed.

One small niggle is that I couldn't build with ninja - the `error: invalid argument '-std=c++11' not allowed with 'C'` problem was triggered while I did not need to set anything for cmake.

```
# Record of working recipe. Picks up some system libraries, not a comprehensive/system independent recipe.
spack load cmake@3.19.1
spack load boost@1.76.0
spack load python@3.8.9%gcc@11.1.0 # Includes test python deps such as numpy
cmake -DCMAKE_C_COMPILER=hipcc -DCMAKE_CXX_COMPILER=hipcc -DQMC_MPI=0 -DQMC_CUDA=1 -DQMC_CUDA2HIP=1 -DQMC_NIO_MAX_SIZE=64 -DQMC_DATA=/scratch/pk7/QMC_DATA_FULL ..
```

**Comment by ye-luo** [comment]
I have no difficult using
```
cmake -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DQMC_MPI=0 -DQMC_CUDA=1 -DQMC_CUDA2HIP=1 ..
cmake -DCMAKE_C_COMPILER=gcc -DCMAKE_CXX_COMPILER=g++ -DQMC_MPI=0 -DQMC_CUDA=1 -DQMC_CUDA2HIP=1 ..
```
hipcc is picked up by CMake to build cu files.

On the C11, it is a simple change to modernize our CMake.

**Comment by prckent** [comment]
Looking at the diamond 2x1x1 case, the short tests as well as the deterministic tests also fail. The short tests are just about long enough for reliable statistics. From the failures we can see that the wavefunction is wrong - all the energy components that are checked are wrong indicating that we get the wrong probability distribution in the Monte Carlo. Since the jastrow function and the terms in the energy evaluation appear to be correct in other tests, we can infer that some part of the spline handling (load, process, evaluation) is wrong in the 2x1x1 case. If the electron-electron energy is wrong in this test we know that the value^2 of the wavefunction is wrong somewhere. If this was correct but the kinetic energy was wrong we would know that the second derivative of the wavefunction is wrong.

```
391/546 Test #355: short-diamondC_2x1x1_pp-vmc_sdj-1-16-totenergy ...........................***Failed    0.08 sec
Tests for series 0
  Testing quantity: LocalEnergy
    reference mean value     : -21.66829300
    reference error bar      :   0.00700000
    computed  mean value     : -27.82064095
    computed  error bar      :   0.01322219
    pass tolerance           :   0.02100000  (  3.00000000 sigma)
    deviation from reference :  -6.15234795  (-878.90685054 sigma)
    error bar of deviation   :   0.01496082
    significance probability :   1.00000000  (gaussian statistics)
    status of this test      :   fail

Test status: fail

392/546 Test #354: short-diamondC_2x1x1_pp-vmc_sdj-1-16-kinetic .............................***Failed    0.08 sec
Tests for series 0
  Testing quantity: Kinetic
    reference mean value     :  20.06515100
    reference error bar      :   0.04400000
    computed  mean value     :   9.48942833
    computed  error bar      :   0.03527041
    pass tolerance           :   0.13200000  (  3.00000000 sigma)
    deviation from reference : -10.57572267  (-240.35733348 sigma)
    error bar of deviation   :   0.05639151
    significance probability :   1.00000000  (gaussian statistics)
    status of this test      :   fail

Test status: fail393/546 Test #356: short-diamondC_2x1x1_pp-vmc_sdj-1-16-eeenergy ............................***Failed    0.08 sec
Tests for series 0
  Testing quantity: ElecElec
    reference mean value     :  -5.42856100
    reference error bar      :   0.01200000
    computed  mean value     :  -6.14289649
    computed  error bar      :   0.01556417
    pass tolerance           :   0.03600000  (  3.00000000 sigma)
    deviation from reference :  -0.71433549  (-59.52795725 sigma)
    error bar of deviation   :   0.01965308
    significance probability :   1.00000000  (gaussian statistics)
    status of this test      :   fail

Test status: fail
```

Also in my case the Hamiltonian unit tests are failing and somehow have the test_ion_derivs check enabled. This should only happen when QMC_CUDA is not set, so there is some cmake issue here.

**Review by prckent — CHANGES_REQUESTED** [review]
This does not build with QMC_COMPLEX=1 . Please can you fix? It indicates that a type is wrong somewhere, and might even be why the real build has a few problems.

```
[ 20%] Building HIPCC object src/QMCWaveFunctions/CMakeFiles/qmcwfs_cuda.dir/detail/CUDA_legacy/qmcwfs_cuda_generated_determinant_update.cu.o
/home/pk7/projects/qmc/git_QMCPACK_prckent/qmcpack/src/QMCWaveFunctions/detail/CUDA_legacy/determinant_update.cu:192:14: error: no template named 'uninitialized_array'
  __shared__ uninitialized_array<T, BS> delta;
             ^
/home/pk7/projects/qmc/git_QMCPACK_prckent/qmcpack/src/QMCWaveFunctions/detail/CUDA_legacy/determinant_update.cu:254:14: error: no template named 'uninitialized_array'
  __shared__ uninitialized_array<T, BS> Ainv_colk_shared;
             ^
/home/pk7/projects/qmc/git_QMCPACK_prckent/qmcpack/src/QMCWaveFunctions/detail/CUDA_legacy/determinant_update.cu:303:14: error: no template named 'uninitialized_array'
  __shared__ uninitialized_array<T, BS> Ainv_colk_shared;
             ^
/home/pk7/projects/qmc/git_QMCPACK_prckent/qmcpack/src/QMCWaveFunctions/detail/CUDA_legacy/determinant_update.cu:468:26: error: use of undeclared identifier 'thrust'
  update_inverse_kernel1<thrust::complex<float>, BS1><<<dimGrid1, dimBlock1>>>((thrust::complex<float>**)data,
```

**Comment by jakurzak** [comment]
The problem with QMC_COMPLEX was only due to header files and is fixed now.

**Comment by prckent** [comment]
I ran full precision tests with the complex and real version (QMC_COMPLEX=0/1 and QMC_MIXED_PRECISION=0).

- diamondC_2x1x1 is broken with real but not complex. As discussed above, something to do with spline handling is most likely be to blame.
- The delayed update code path appears to be broken, e.g. short-diamondC_2x1x1_pp-delayed_update-vmc_sdj-* and deterministic-diamondC_2x1x1_pp-delayed_update-vmc_sdj-* fail with complex (as well as real, but those runs have extra issues). Interestingly in the real build short-NiO_a4_e48_pp-delayed_update-dmc_sdj-1-16 aborts and there are NaNs reported in the VMC section output.

This seems close enough to be working that I think we should merge and then setup a nightly build to track progress.

**Comment by prckent** [comment]
@jakurzak Planning to merge this shortly unless @ye-luo has any final comments. Some problems with test machine reliability while I was working on this yesterday. I hope to get nightly builds working.

**Review by ye-luo — APPROVED** [review]

**Comment by prckent** [comment]
Interestingly on nitrogen (~up to date RHEL8+Rome+RadeonVII+spack installed libraries) the following now causes a system lockup:
```
cmake -DCMAKE_C_COMPILER=gcc -DCMAKE_CXX_COMPILER=g++ -DQMC_MPI=0 -DQMC_CUDA=1 -DQMC_CUDA2HIP=1 -DQMC_NIO_MAX_SIZE=64 -DQMC_DATA=/scratch/pk7/QMC_DATA_FULL -DQMC_COMPLEX=0 -DQMC_MIXED_PRECISION=0 ../qmcpack/
```
```
-- Setting Boost_INCLUDE_DIRS=/home/pk7/apps/spack/opt/spack/linux-rhel8-x86_64/gcc-11.1.0/boost-1.76.0-nqdherqzxh3ei3fexdduemkbpgak5yst/include
-- Found HIP: /opt/rocm-4.2.0/hip (found version "4.2.21155-37cb3a34") 
-- ROCclr at /opt/rocm/lib/cmake/rocclr
-- hip::amdhip64 is SHARED_LIBRARY
-- CUDA NVTX APIs disabled
-- VTune ittnotify APIs disabled
-- QMC_DATA is /scratch/pk7/QMC_DATA_FULL
Project C_FLAGS:  -fopenmp -finline-limit=1000 -fstrict-aliasing -funroll-all-loops -Wno-deprecated -Werror=vla -march=native -O3 -DNDEBUG -fomit-frame-pointer -ffast-math
Project CXX_FLAGS:  -fopenmp -finline-limit=1000 -fstrict-aliasing -funroll-all-loops -D__forceinline=inline -Wno-deprecated -Wvla -Wcomment -Wmisleading-indentation -Wmaybe-uninitialized -Wuninitialized -Wreorder -Wno-unknown-pragmas -Wno-sign-compare -march=native -O3 -DNDEBUG -fomit-frame-pointer -ffast-math
Project INCLUDE_DIRECTORIES: /home/pk7/projects/qmc/git_QMCPACK_prckent/qmcpack/src;/home/pk7/projects/qmc/git_QMCPACK_prckent/build_gcc_hip_full/src
Project EXE_LINKER_FLAGS:  
Project SHARED_LINKER_FLAGS:  
-- Ready to build qmcpack
sed supports -E
Git branch: pr/jakurzak/3165
Git commit hash: e8c5d92703adfb2e87dbfa0d11f4aa4d5e86f2e9
-- Adding integration tests for QMCPACK
Set QMC_VERBOSE_CONFIGURATION=1 for details of which tests will be enabled.
Adding system tests for QMCPACK
TEST_MAX_PROCS was unset. Set to 16
/scratch/pk7/QMCPACK_CI_BUILDS_DO_NOT_REMOVE/build_gccnew_pyscf/pyscf/pyscf/lib/misc.py:47: H5pyDeprecationWarning: Using default_file_mode other than 'r' is deprecated. Pass the mode to h5py.File() instead.
  h5py.get_config().default_file_mode = 'a'
Python dependencies met. Adding PySCF workflow tests
client_loop: send disconnect: Broken pipe
```

**Comment by ye-luo** [comment]
I'm stuck on spock at
```
-- Generating dependency file: /ccs/home/yeluo/opt/qmcpack/build_spock_gcc_hip_real/src/Platforms/CUDA_legacy/CMakeFiles/platform_cuda_legacy.dir//platform_cuda_legacy_generated_cuda_inverse.cu.o.depend.pre
/sw/spock/spack-envs/views/rocm-4.1.0/hip/bin/hipcc -M /ccs/home/yeluo/opt/qmcpack/src/Platforms/CUDA_legacy/cuda_inverse.cu -o /ccs/home/yeluo/opt/qmcpack/build_spock_gcc_hip_real/src/Platforms/CUDA_legacy/CMakeFiles/platform_cuda_legacy.dir//platform_cuda_legacy_generated_cuda_inverse.cu.o.depend.pre -fPIC -fopenmp --amdgpu-target=gfx906,gfx908 --gpu-max-threads-per-block=256 -Wno-vla -Wno-deprecated-declarations -Wno-unused-command-line-argument -Drestrict=__restrict__ -D__HIP_PLATFORM_HCC__ -DHAVE_CONFIG_H -DUSE_REAL_STRUCT_FACTOR -DADD_ -I/autofs/nccs-svm1_sw/spock/spack-envs/views/rocm-4.1.0/hip/include -I/sw/spock/spack-envs/views/rocm-4.1.0/llvm/lib/clang/12.0.0/include/.. -I/autofs/nccs-svm1_sw/spock/spack-envs/views/rocm-4.1.0/hip/include -I/sw/spock/spack-envs/views/rocm-4.1.0/include -I/ccs/home/yeluo/opt/qmcpack/src -I/ccs/home/yeluo/opt/qmcpack/build_spock_gcc_hip_real/src -I/ccs/home/yeluo/opt/qmcpack/external_codes/boost_multi -I/autofs/nccs-svm1_sw/spock/spack-envs/base/opt/linux-sles15-x86_64/gcc-7.5.0/hipblas-4.1.0-mskwrvzyzw47xa4a4xps2ohiotcfynxt/include -I/autofs/nccs-svm1_sw/spock/spack-envs/base/opt/linux-sles15-x86_64/gcc-7.5.0/hipblas-4.1.0-mskwrvzyzw47xa4a4xps2ohiotcfynxt/include -I/autofs/nccs-svm1_sw/spock/spack-envs/views/rocm-4.1.0/hip/include /sw/spock/spack-envs/views/rocm-4.1.0/llvm/lib/clang/12.0.0/include/.. /autofs/nccs-svm1_sw/spock/spack-envs/views/rocm-4.1.0/hip/include /sw/spock/spack-envs/views/rocm-4.1.0/include -I/ccs/home/yeluo/opt/qmcpack/src -I/ccs/home/yeluo/opt/qmcpack/build_spock_gcc_hip_real/src -I/ccs/home/yeluo/opt/qmcpack/external_codes/boost_multi
clang-12: error: cannot specify -o when generating multiple output files
CMake Error at platform_cuda_legacy_generated_cuda_inverse.cu.o.cmake:142 (message):
  Error generating
  /ccs/home/yeluo/opt/qmcpack/build_spock_gcc_hip_real/src/Platforms/CUDA_legacy/CMakeFiles/platform_cuda_legacy.dir//./platform_cuda_legacy_generated_cuda_inverse.cu.o
```
when using rocm 4.1 + cmake 3.17/3.20. 4.2 is not available on spock.
My cmake hack to workaround include path issue.
```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index bc00b8972..a9a00661d 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -748,8 +748,9 @@ IF(QMC_CUDA OR ENABLE_CUDA)
     LIST(APPEND CMAKE_MODULE_PATH ${ROCM_ROOT}/hip/cmake)
     FIND_PACKAGE(HIP REQUIRED)
     FIND_PACKAGE(hipblas REQUIRED)
-    LINK_LIBRARIES(amdhip64 roc::hipblas)
-    ADD_DEFINITIONS(-D__HIP_PLATFORM_AMD__)
+    INCLUDE_DIRECTORIES("$<TARGET_PROPERTY:hip::device,INTERFACE_INCLUDE_DIRECTORIES>")
+    LINK_LIBRARIES(roc::hipblas)
+    ADD_DEFINITIONS(-D__HIP_PLATFORM_HCC__)
     # architecture flags
     SET(HIP_ARCH "gfx906,gfx908" CACHE STRING "HIP architecture gfxXXX")
     LIST(APPEND HIP_HIPCC_FLAGS "-fPIC -fopenmp")
```

**Review by prckent — APPROVED** [review]
Due to the above mentioned problems we aren't going to be able to build this for some time. However since it is benign to the other builds I think we can merge now to avoid risk of further divergence and then revisit HIP support when we have suitable test machines.

**Pull request merged**